### PR TITLE
Fix toctou in create

### LINKED ISSUE #1401: ignore_conflict option for create() is not used anywhere
In Kinto or in any of the plugins I checked.

I propose we get rid of it

Ref #1122

### PR DESCRIPTION
**Pull request by glasserc**
Fixes #1401, #1376.

- (n/a) Add documentation.
- (n/a) Add tests.
- [x] Add a changelog entry.
- (n/a) Add your name in the contributors file.
- (n/a) If you changed the HTTP API, update the [API_VERSION](https://github.com/Kinto/kinto/blob/master/kinto/__init__.py#L15) constant and add an API changelog entry [in the docs](https://github.com/Kinto/kinto/blob/master/docs/api/index.rst)
- (n/a) If you added a new configuration setting, update the [`kinto.tpl`](https://github.com/Kinto/kinto/blob/master/kinto/config/kinto.tpl) file with it.

### CODE REVIEW

File: kinto/core/storage/postgresql/__init__.py [hunk 1]
```diff
@@ -252,57 +252,47 @@ def collection_timestamp(self, collection_id, parent_id, auth=None):
...
-        RETURNING id, as_epoch(last_modified) AS last_modified;
-        """
-
-        safe_holders = {}
-
-        if ignore_conflict:
-            # If we ignore conflict, then we do not touch the existing data.
-            # Unless if the conflict comes from a tombstone.
-            safe_holders['on_conflict'] = """
+        WITH ins AS (
```
Comment (leplatrem) [inline kinto/core/storage/postgresql/__init__.py]: nit: `ins` → `create_record`

File: kinto/core/storage/postgresql/__init__.py [hunk 2]
```diff
@@ -252,57 +252,47 @@ def collection_timestamp(self, collection_id, parent_id, auth=None):
...
         query_record.pop(modified_field, None)
 
+        # If there is a record in the table and it is deleted = TRUE,
+        # we want to replace it. Otherwise, we want to do nothing and
+        # throw a UnicityError. Per
+        # https://stackoverflow.com/questions/15939902/is-select-or-insert-in-a-function-prone-to-race-conditions/15950324#15950324
+        # a WHERE clause in the DO UPDATE will lock the conflicting
+        # row whether it is true or not, so the subsequent SELECT is
+        # safe. We add a dummy "inserted" field to know whether we
+        # need to throw or not.
```
Comment (leplatrem) [inline kinto/core/storage/postgresql/__init__.py]: Why `dummy`?

File: kinto/core/storage/postgresql/__init__.py [hunk 3]
```diff
@@ -252,57 +252,47 @@ def collection_timestamp(self, collection_id, parent_id, auth=None):
...
-            """
+            WHERE records.deleted = TRUE
+            RETURNING id, data, last_modified
+        )
+        SELECT id, data, as_epoch(last_modified) AS last_modified, TRUE AS inserted FROM ins
+        UNION ALL
+        SELECT id, data, as_epoch(last_modified) AS last_modified, FALSE AS inserted FROM records
+        WHERE id = :object_id AND parent_id = :parent_id AND collection_id = :collection_id
+        LIMIT 1;
+        """
```
Comment (leplatrem) [inline kinto/core/storage/postgresql/__init__.py]: 👍

File: kinto/core/storage/postgresql/__init__.py [hunk 4]
```diff
@@ -312,6 +302,12 @@ def create(self, collection_id, parent_id, record, id_generator=None,
             result = conn.execute(query % safe_holders, placeholders)
             inserted = result.fetchone()
 
+        if not inserted['inserted']:
+            record = inserted['data']
+            record[id_field] = inserted['id']
+            record[modified_field] = inserted['last_modified']
+            raise exceptions.UnicityError(id_field, record)
```
Comment (leplatrem) [inline kinto/core/storage/postgresql/__init__.py]: 👍

**Comment by leplatrem** [comment]
I don't know if we want to tackle this here (or if we track it in another issue), but in the `resource/__init__` there is also a very ugly `storage.get() + if + storage.create()` code, whereas we could probably catch `UnicityError`

**Comment by glasserc** [comment]
I'm not actually sure how to fix the TOCTOU at the resource level. The reason we do a get first is to enforce 412 concurrency headers. Would we have to do a SELECT FOR UPDATE? If so, that would probably require another break in compatibility with the storage backend, in which case yes maybe we should do it in this PR.

**Comment by glasserc** [comment]
N.B. the TOCTOU bug at the resource level affects `put`, `patch`, `delete`, and `collection_post`.

Another approach might be to have the `INSERT`/`UPDATE`/`DELETE` queries return any replaced rows, and do the 412 afterwards, rolling back if necessary??

**Comment by glasserc** [comment]
Or another option is to push the 412 conditions into the SQL layer somehow. This would also be a breaking change...

**Comment by glasserc** [comment]
I'm going to merge this for now and punt on the TOCTOU at the resource level, which is now #1407. I'll cut a release and at least we can cut down on the Sentry errors..

**Pull request merged**